google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/RetryInfoTest.java

```diff
@@ -0,0 +1,668 @@
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class RetryInfoTest {
+
+  @Rule public GrpcServerRule serverRule = new GrpcServerRule();
+
+  private FakeBigtableService service;
+  private BigtableDataClient client;
+  private BigtableDataSettings.Builder settings;
+
+  private AtomicInteger attemptCounter = new AtomicInteger();
+  private Duration delay = Duration.newBuilder().setSeconds(1).setNanos(0).build();
+
+  @Before
+  public void setUp() throws IOException {
+    service = new FakeBigtableService();
+    serverRule.getServiceRegistry().addService(service);
+
+    settings =
+        BigtableDataSettings.newBuilder()
+            .setProjectId("fake-project")
+            .setInstanceId("fake-instance")
+            .setCredentialsProvider(NoCredentialsProvider.create());
+
+    settings
+        .stubSettings()
+        .setTransportChannelProvider(
+            FixedTransportChannelProvider.create(
+                GrpcTransportChannel.create(serverRule.getChannel())))
+        // channel priming doesn't work with FixedTransportChannelProvider. Disable it for the test
+        .setRefreshingChannel(false)
+        .build();
+
+    this.client = BigtableDataClient.create(settings.build());
+  }
+
+  @Test
+  public void testReadRow() {
+    createRetryableExceptionWithDelay(delay);
+    Stopwatch stopwatch = Stopwatch.createStarted();
+    client.readRow("table", "row");
+    stopwatch.stop();
+
+    assertThat(attemptCounter.get()).isEqualTo(2);
+    assertThat(stopwatch.elapsed()).isAtLeast(java.time.Duration.ofSeconds(delay.getSeconds()));
+  }
+
+  @Test
+  public void testReadRowNonRetryableErrorWithRetryInfo() {
+    createNonRetryableExceptionWithDelay(delay);
+    Stopwatch stopwatch = Stopwatch.createStarted();
+    client.readRow("table", "row");
+    stopwatch.stop();
+
+    assertThat(attemptCounter.get()).isEqualTo(2);
+    assertThat(stopwatch.elapsed()).isAtLeast(java.time.Duration.ofSeconds(delay.getSeconds()));
+  }
+
+  @Test
+  public void testReadRowDisableRetryInfo() throws IOException {
+    settings.stubSettings().setEnableRetryInfo(false);
+
+    try (BigtableDataClient client = BigtableDataClient.create(settings.build())) {
+      createRetryableExceptionWithDelay(delay);
+      Stopwatch stopwatch = Stopwatch.createStarted();
+      client.readRow("table", "row");
+      stopwatch.stop();
+
+      assertThat(attemptCounter.get()).isEqualTo(2);
+      assertThat(stopwatch.elapsed()).isLessThan(java.time.Duration.ofSeconds(delay.getSeconds()));
+
+      attemptCounter.set(0);
+      ApiException exception = createNonRetryableExceptionWithDelay(delay);
+      try {
+        client.readRow("table", "row");
+      } catch (ApiException e) {
+        assertThat(e.getStatusCode()).isEqualTo(exception.getStatusCode());
+      }
+      assertThat(attemptCounter.get()).isEqualTo(1);
+    }
+  }
```

Comment:
there is a lot of repetition for each method here. Unless I'm mistaken can you just write a single test that takes a Runnable and then just pass the method invocation there?